src/cargo/ops/cargo_package.rs

```diff
@@ -157,7 +164,74 @@ fn create_package(
     return Ok(dst);
 }
 
-pub fn package(ws: &Workspace<'_>, opts: &PackageOpts<'_>) -> CargoResult<Option<Vec<FileLock>>> {
+/// Determine which registry the packages are for.
+///
+/// The registry only affects the built packages if there are dependencies within the
+/// packages that we're packaging: if we're packaging foo-bin and foo-lib, and foo-bin
+/// depends on foo-lib, then the foo-lib entry in foo-bin's lockfile will depend on the
+/// registry that we're building packages for.
+fn infer_registry(
+    gctx: &GlobalContext,
+    pkgs: &[&Package],
+    reg_or_index: Option<RegistryOrIndex>,
+) -> CargoResult<SourceId> {
+    let publish_registry = match reg_or_index.as_ref() {
+        Some(RegistryOrIndex::Registry(registry)) => Some(registry.clone()),
+        _ => {
+            // If no registry is specified in the command, but all of the packages
+            // to publish have the same, unique allowed registry, push to that one.
+            pkgs[0].publish().as_ref().and_then(|allowed| {
+                if allowed.len() == 1
+                    && pkgs[1..]
+                        .iter()
+                        .all(|p| p.publish().as_ref() == Some(allowed))
+                {
+                    Some(allowed[0].clone())
+                } else {
+                    None
+                }
+            })
+        }
+    };
```

Comment:
nit: this is too complex of an expression to use combinators.  It particularly is confusing as `pkgs[1..]` is read inside of the combinator.  The general expectation is that the combinator is checking whats inside of itself, with inputs.